zcash_client_backend/src/keys.rs

```diff
@@ -507,19 +503,38 @@ impl UnifiedFullViewingKey {
         self.orchard.as_ref()
     }
 
-    /// Attempts to derive the Unified Address for the given diversifier index.
+    /// Attempts to derive the Unified Address for the given diversifier index and
+    /// receiver types.
     ///
     /// Returns `None` if the specified index does not produce a valid diversifier.
-    // TODO: Allow filtering down by receiver types?
-    pub fn address(&self, j: DiversifierIndex) -> Option<UnifiedAddress> {
-        let sapling = if let Some(extfvk) = self.sapling.as_ref() {
+    pub fn address(
+        &self,
+        j: DiversifierIndex,
+        receiver_types: &[Typecode],
+    ) -> Option<UnifiedAddress> {
+        let orchard_j = orchard::keys::DiversifierIndex::from(j.as_bytes().clone());
+        let orchard = self
+            .orchard
+            .as_ref()
+            .filter(|_| receiver_types.contains(&Typecode::Orchard))
+            .map(|ofvk| ofvk.address_at(orchard_j, Scope::External));
+
+        let sapling = if let Some(extfvk) = self
+            .sapling
+            .as_ref()
+            .filter(|_| receiver_types.contains(&Typecode::Sapling))
+        {
             Some(extfvk.address(j)?)
         } else {
             None
         };
 
         #[cfg(feature = "transparent-inputs")]
-        let transparent = if let Some(tfvk) = self.transparent.as_ref() {
+        let transparent = if let Some(tfvk) = self
+            .transparent
+            .as_ref()
+            .filter(|_| receiver_types.contains(&Typecode::P2pkh))
+        {
             match to_transparent_child_index(j) {
                 Some(transparent_j) => match tfvk
                     .derive_external_ivk()
```

Comment:
Re: the `return None`, are we sure we want to return from the whole method if we can't derive the transparent address, rather than just omitting it from the UA? (If we want the latter then just use `Result::ok()` rather than the inner match.)